File: Server/src/main/java/com/mancersoft/litevpnserver/transport/WebSocketTransport.java
package com.mancersoft.litevpnserver.transport;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.mancersoft.litevpnserver.VpnManager;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import trikita.log.Log;

import java.net.URI;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

import static com.mancersoft.litevpnserver.VpnManager.TAG;

public class WebSocketTransport implements IVpnTransport {

    private static final String SERVER_URI = "wss://cloud.achex.ca";

    private final String mServerId;
    private URI serverUri;
    private MessageListener mMessageListener;
    private WSClient mWebSocketClient;

    public WebSocketTransport(String serverId) {
        mServerId = serverId;
        try {
            serverUri = new URI(SERVER_URI);
        } catch (Exception e) {
            Log.e(TAG, "WebSocketTransport URI error", e);
        }
    }

    @Override
    public boolean isReliable() {
        return true;
    }

    @Override
    public CompletableFuture<Boolean> start() {
        stop();
        mWebSocketClient = new WSClient(serverUri);
        return mWebSocketClient.connectAsync();
    }

    @Override
    public void stop() {
        if (mWebSocketClient != null) {
            mWebSocketClient.disconnect();
        }
    }

    private void restart() {
        Log.d(TAG, "WebSocketTransport restarting...");
        start();
    }

    @Override
    public void sendAsync(Packet packet) {
        var encoded = Base64.getEncoder().encode(ByteBuffer.wrap(packet.getData(), 0, packet.getLength()));
        String dataString = new String(encoded.array(), StandardCharsets.ISO_8859_1);
        var jObj = new JsonObject();
        jObj.addProperty("to", packet.getDestination().toString());
        jObj.addProperty("data", dataString);
        mWebSocketClient.sendAsync(jObj.toString());
    }

    @Override
    public void setOnMessageListener(MessageListener messageListener) {
        mMessageListener = messageListener;
    }

    private class WSClient extends WebSocketClient {

        private CompletableFuture<Boolean> mConnectResult;

        public WSClient(URI serverUri) {
            super(serverUri);
        }

        public CompletableFuture<Boolean> connectAsync() {
            mConnectResult = new CompletableFuture<>();
            super.connect();
            return mConnectResult;
        }

        public void disconnect() {
            super.close();
        }

        @Override
        public void onOpen(ServerHandshake handshakeData) {
            Log.d(TAG, "WebSocketTransport onOpen");
            var jObj = new JsonObject();
            jObj.addProperty("auth", mServerId);
            jObj.addProperty("passwd", UUID.randomUUID().toString());
            sendAsync(jObj.toString());
        }

        public void onMessage(String message) {
            try {

                var jObj = JsonParser.parseString(message).getAsJsonObject();
                var auth = jObj.get("auth");
                if (auth != null) {
                    mConnectResult.complete(auth.getAsString().toLowerCase().equals("ok"));
                    return;
                }

                String dataString = jObj.get("data").getAsString();
                var packet = new Packet();
                byte[] data = Base64.getDecoder().decode(dataString);
                if (data.length > VpnManager.MAX_PACKET_SIZE) {
                    throw new Exception("Oversized packet received");
                }

                packet.setData(data);
                packet.setLength(data.length);
                packet.setSource(jObj.get("FROM").getAsString());
                if (mMessageListener != null) {
                    mMessageListener.onMessage(packet);
                }
            } catch (Exception e) {
                Log.e(TAG, "WebSocketTransport onMessage error", e);
            }
        }

        @Override
        public void onClose(int code, String reason, boolean remote) {
            Log.d(TAG, "WebSocketTransport onClose reason: " + reason);
            restart();
        }

        @Override
        public void onError(Exception e) {
            Log.e(TAG, "WebSocketTransport WSClient error", e);
            if (mConnectResult != null) {
                mConnectResult.complete(false);
            }
        }

        public void sendAsync(String message) {
            super.send(message);
        }
    }
}

File: Server/src/main/kotlin/com/mancersoft/litevpnserver/transport/WebSocketTransport.kt
package com.mancersoft.litevpnserver.transport

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.mancersoft.litevpnserver.VpnManager
import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import trikita.log.Log
import java.net.URI
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.CompletableFuture

private const val SERVER_URI = "wss://cloud.achex.ca"

class WebSocketTransport(private val mServerId: String) : IVpnTransport {

    private lateinit var serverUri: URI
    private var mMessageListener: ((Packet) -> Unit)? = null
    private var mWebSocketClient: WSClient? = null
    override val isReliable: Boolean
        get() = true

    init {
        try {
            serverUri = URI(SERVER_URI)
        } catch (e: Exception) {
            Log.e(VpnManager.TAG, "WebSocketTransport URI error", e)
        }
    }

    override fun start(): CompletableFuture<Boolean> {
        stop()
        mWebSocketClient = WSClient(serverUri)
        return mWebSocketClient!!.connectAsync()
    }

    override fun stop() {
        mWebSocketClient?.disconnect()
    }

    private fun restart() {
        Log.d(VpnManager.TAG, "WebSocketTransport restarting...")
        start()
    }

    override fun sendAsync(packet: Packet) {
        val encoded = Base64.getEncoder().encode(ByteBuffer.wrap(packet.data, 0, packet.length))
        val dataString = String(encoded.array(), StandardCharsets.ISO_8859_1)
        val jObj = JsonObject()
        jObj.addProperty("to", packet.destination.toString())
        jObj.addProperty("data", dataString)
        mWebSocketClient!!.sendAsync(jObj.toString())
    }

    override fun setOnMessageListener(messageListener: ((Packet) -> Unit)?) {
        mMessageListener = messageListener
    }

    private inner class WSClient(serverUri: URI?) : WebSocketClient(serverUri) {

        private var mConnectResult: CompletableFuture<Boolean>? = null

        fun connectAsync(): CompletableFuture<Boolean> {
            mConnectResult = CompletableFuture()
            super.connect()
            return mConnectResult!!
        }

        fun disconnect() {
            super.close()
        }

        override fun onOpen(handshakeData: ServerHandshake) {
            Log.d(VpnManager.TAG, "WebSocketTransport onOpen")
            val jObj = JsonObject()
            jObj.addProperty("auth", mServerId)
            jObj.addProperty("passwd", UUID.randomUUID().toString())
            sendAsync(jObj.toString())
        }

        override fun onMessage(message: String) {
            try {
                val jObj = JsonParser.parseString(message).asJsonObject
                val auth = jObj["auth"]
                if (auth != null) {
                    mConnectResult?.complete(auth.asString.toLowerCase() == "ok")
                    return
                }

                val error = jObj["error"];
                if (error != null) {
                    Log.d("WebSocketTransport onMessage Achex error: $error")
                    return
                }

                val dataString = jObj["data"].asString
                val packet = Packet()
                val data = Base64.getDecoder().decode(dataString)
                if (data.size > VpnManager.MAX_PACKET_SIZE) {
                    throw Exception("Oversized packet received")
                }
                packet.data = data
                packet.length = data.size
                packet.source = jObj["FROM"].asString
                mMessageListener?.invoke(packet)
            } catch (e: Exception) {
                Log.e(VpnManager.TAG, "WebSocketTransport onMessage error;" +
                        "\nmessage:\n$message", e)
            }
        }

        override fun onClose(code: Int, reason: String, remote: Boolean) {
            Log.d(VpnManager.TAG, "WebSocketTransport onClose reason: $reason")
            restart()
        }

        override fun onError(e: Exception) {
            Log.e(VpnManager.TAG, "WebSocketTransport WSClient error", e)
            mConnectResult?.complete(false)
        }

        fun sendAsync(message: String?) {
            super.send(message)
        }
    }
}